torchci/lib/jobUtils.ts

```diff
@@ -144,3 +151,70 @@ export function isSameFailure(
     isEqual(jobA.failure_captures, jobB.failure_captures)
   );
 }
+
+export async function removeCancelledJobAfterRetry<T extends BasicJobData>(
+  jobs: T[]
+): Promise<T[]> {
+  // When a worlflow is manually cancelled and retried, the leftover cancel signals from
+  // the previous workflow run are poluting HUD and Dr.CI. For example, the pull request
+  // https://hud.pytorch.org/pytorch/pytorch/pull/107339 had many cancelled binary build
+  // jobs showing up as new failures after the workflow had been retried successfully.
+  //
+  // The issue here is that the cancelled job name is not the same as the successfully
+  // retried one, for example manywheel-py3_10-cuda11_8-test (cancel) was retried as
+  // manywheel-py3_10-cuda11_8-test / test (success). As their names look different,
+  // HUD and Dr.CI treat them incorrectly as two different jobs and mark the cancelled
+  // one as a failure.
+  //
+  // So the fix here is to check if a cancelled job has been retried successfully and
+  // keep or remove it from the list accordingly.
+  const trie: TrieSearch<T> = new TrieSearch<T>("name", {
+    splitOnRegEx: /\s\/\s/g,
+  });
+  trie.addAll(jobs);
+
+  const processedJobName: Set<string> = new Set<string>();
+  const filteredJobs: T[] = [];
+
+  for (const job of jobs) {
+    if (job.name === undefined) {
+      continue;
+    }
+
+    let currentMatch: T | undefined = undefined;
+    let currentLatestTimestamp = dayjs(0);
+
+    const matches = trie.search(job.name);
+    if (matches.length <= 1) {
+      // If there is zero or one match, keep the job as it is as this is no retry
+      currentMatch = job;
+    } else {
+      // NB: Default to the latest successful job. This is needed because the event
+      // time from GitHub does not guarantee strict chronological order. A quick
+      // retry event could have a timestamp few seconds earlier than the original
+      // job. We are getting the last successful job here (if any)
+      currentMatch = _.find(matches, (match: T) => isSuccessJob(match));
```

Comment:
Why default to the latest successful job instead of simply the latest job, and then seeing if it was successful or not?

This way we'd always show the most recent retry, even in case of failures

Reply:
Well, the issue here is that the latest job according to GitHub timestamp is the one that gets cancelled.  Take jobs from `1f762dfc92b46323950c3e6e95d99d9687741451` as an example, here is what happens for in real time:

1. The `Labeler` job with workflow id `5949201202` at https://github.com/pytorch/pytorch/actions/runs/5949201202 started and was cancelled by the workflow concurrency rule.  Its timestamp was `2023-08-23T08:57:20.619274Z`
2. The successful retried job with id `5949201205` (you can see that the id is bigger) at https://github.com/pytorch/pytorch/actions/runs/5949201205 had the timestamp `2023-08-23T08:57:20.499395Z`

They happened in this order, but you can see that the workflow timestamp of the retry `08:57:20.499395Z` recorded by GitHub was a split second earlier than the cancelled run `08:57:20.619274Z`.

So simply getting the latest job here would wrongly surface the wrong result.  The fix I have here is:

1. Get the timestamp of the latest successful job as the base
2. Check if there are any jobs after that, but at least one minute later.  This is the meaning of `timestamp.isAfter(currentLatestTimestamp, "minute")`
3. Use the conclusion from the latest job found by the second step

Reply:
I don't think the rerun happened after the job succeed. The job was just cancelled and retried so quickly within second, that GitHub timestamps were in the wrong order.  This could happen in a distributed system I think